crates/neon/src/types_impl/extract/mod.rs

```diff
@@ -99,24 +99,27 @@
 //! Note well, in this example, type annotations are not required on the tuple because
 //! Rust is able to infer it from the type arguments on `add` and `concat`.
 
-use std::{fmt, marker::PhantomData};
-
 use crate::{
     context::{Context, Cx, FunctionContext},
     handle::Handle,
-    result::{JsResult, NeonResult, ResultExt},
+    result::{JsResult, NeonResult},
     types::{JsValue, Value},
 };
 
-pub use self::{error::Error, with::With};
+pub use self::{
+    error::{Error, TypeExpected},
+    with::With,
+};
 
 #[cfg(feature = "serde")]
 #[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
 pub use self::json::Json;
 
-mod error;
 #[cfg(feature = "serde")]
-mod json;
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+pub mod json;
```

Comment:
This is public now. It felt better to expose the JSON error wrapper as `json::Error` than to move it up a level as `JsonError`.

I kept the `Json` extractor double exposed at `extract::Json` so that it can live next to the other extractors. Most users won't need to dig into the deeper module for the error.